src/main/scala/mesosphere/marathon/storage/repository/GcActor.scala

```diff
@@ -291,65 +294,68 @@ private[storage] trait ScanBehavior[K, C, S] extends StrictLogging { this: FSM[S
       podVersionsInUse.map { case (id, pods) => id -> pods.to[Set] }(collection.breakOut)
     }
 
-    def rootsInUse(): Future[Seq[StoredGroup]] = {
-      Future.sequence {
-        storedPlans.flatMap(plan =>
-          Seq(
-            groupRepository.lazyRootVersion(plan.originalVersion),
-            groupRepository.lazyRootVersion(plan.targetVersion))
-        )
-      }
-    }.map(_.flatten)
+    def rootsInUse(): Source[StoredGroup, NotUsed] = {
+      Source(storedPlans)
+        .mapConcat(plan => Seq(plan.originalVersion, plan.targetVersion))
+        .mapAsync(1)(groupRepository.lazyRootVersion)
+        .mapConcat(_.toList)
+    }
 
     def appsExceedingMaxVersions(usedApps: Set[PathId]): Future[Map[PathId, Set[OffsetDateTime]]] = {
-      Future.sequence {
-        usedApps.map { id =>
-          appRepository.versions(id).runWith(Sink.sortedSet).map(id -> _)
-        }
-      }.map(_.filter(_._2.size > maxVersions).toMap)
+      Source(usedApps)
+        .mapAsync(1)(id => appRepository.versions(id).runWith(Sink.sortedSet).map(id -> _))
+        .filter(_._2.size > maxVersions)
+        .runWith(Sink.map)
     }
 
     def podsExceedingMaxVersions(usedPods: Set[PathId]): Future[Map[PathId, Set[OffsetDateTime]]] = {
-      Future.sequence {
-        usedPods.map { id =>
-          podRepository.versions(id).runWith(Sink.sortedSet).map(id -> _)
-        }
-      }.map(_.filter(_._2.size > maxVersions).toMap)
+      Source(usedPods)
+        .mapAsync(1)(id => podRepository.versions(id).runWith(Sink.sortedSet).map(id -> _))
+        .filter(_._2.size > maxVersions)
+        .runWith(Sink.map)
     }
 
-    async { // linter:ignore UnnecessaryElseBranch
-      val inUseRootFuture = rootsInUse()
-      val allAppIdsFuture = appRepository.ids().runWith(Sink.set)
-      val allPodIdsFuture = podRepository.ids().runWith(Sink.set)
-      val allAppIds = await(allAppIdsFuture)
-      val allPodIds = await(allPodIdsFuture)
-      val inUseRoots = await(inUseRootFuture)
-      val usedApps = appsInUse(inUseRoots)
-      val usedPods = podsInUse(inUseRoots)
-      val appsWithTooManyVersions = await(appsExceedingMaxVersions(usedApps.keySet))
-      val podsWithTooManyVersions = await(podsExceedingMaxVersions(usedPods.keySet))
-
-      val appVersionsToDelete = appsWithTooManyVersions.map {
-        case (id, versions) =>
-          val candidateVersions = versions.diff(usedApps.getOrElse(id, SortedSet.empty))
-          id -> candidateVersions.take(versions.size - maxVersions)
-      }
+    val allAppIdsFuture = appRepository.ids().runWith(Sink.set)
+    val allPodIdsFuture = podRepository.ids().runWith(Sink.set)
+
+    rootsInUse()
+      .grouped(scanBatchSize)
+      .mapAsync(1) { inUseRoots =>
```

Comment:
Could you annotate the size of `inUseRoots`? Is this a batch?

Reply:
Yes it is.